### ISSUE TITLE
Future work: WeakRange

### ISSUE THREAD

**Issue by CaseyCarter** [CaseyCarter/cmcstl2#15]
```c++
template <class S, class I>
concept WeakSentinel =
  WeakIterator<I>() && Semiregular<S>() &&
  requires (const I& i, const S& s) {
    { i == s } -> Boolean;
    { i != s } -> Boolean;
    // Axiom: bool(i != s) == !bool(i == s)
  };

template <class R>
  requires WeakIterator<decltype(begin(declval<remove_reference_t<R>&>()))>()
using IteratorType = decltype(begin(declval<remove_reference_t<R>&>()));

template <class R>
  requires WeakSentinel<decltype(end(declval<remove_reference_t<R>&>())), IteratorType<R>>
using SentinelType = decltype(end(declval<remove_reference_t<R>&>()));

template <class R>
concept bool WeakRange =
  requires { typename SentinelType<R>; };

template <class R>
concept bool WeakInputRange =
  WeakRange<R> && WeakInputIterator<IteratorType<R>>();

template <class R, class T>
concept bool WeakOutputRange =
  WeakRange<R> && WeakOutputIterator<IteratorType<R>, T>();

template <class R>
concept bool Range =
  WeakRange<R> && Sentinel<SentinelType<R>, IteratorType<R>>();
```
I postulate that there is no model of `InputIterator` / `OutputIterator` that is not equivalent to some model of `WeakInputIterator` / `WeakOutputIterator`, i.e, `InputIterator` and `OutputIterator` are extraneous and do not increase the expressiveness of the design. If we relax every `Input`/`Output` `Range`/`Iterator` requirement to `Weak` `Input`/`Output` `Range`/`Iteratator` the `Input`/`Output` `Iterator` concepts could be deprecated.

**Comment by ericniebler** [comment]
I don't disagree with this, but it seems that every weak input iterator can be promoted to an input iterator by giving it a trivial op== that returns true, right? Every position besides the last plus one is equivalent from the pov of a single pass algorithm.

**Comment by CaseyCarter** [comment]
Right - that observation is key. A single-pass algorithm cares about two positions: "current" and "end". It would be pointless to compare "current" with itself. It seems silly to me to require input and output iterators to implement `==` when nothing will ever use it. Conversely, `common_iterator<WeakInputIterator, WeakSentinel>` take advantage of that fact to model STL1 `InputIterator` for backwards-compatibility.

I can imagine 10 years from now having to explain to people that `operator==` is useless but they have to implement it because we wanted `InputIterator` and `OutputIterator` to look like their big brother concepts. I understand that it's nice to have things model `Regular` when possible, but `operator==` for what are really weak iterators is a minefield for programming errors: it's useless when used correctly, and UB when used incorrectly. In the near future compilers will tell us we can't compare `WeakIterator`s, but it will be quite a while before we can describe the semantics of `InputIterator` to the degree necessary for a compiler to diagnose `auto tmp = i++; tmp == i;`.

I think we'd be doing future C++ programmers a service if we:
* Eliminate `InputIterator` and `OutputIterator`
* Push the `Iterator` requirements up into `ForwardIterator`
* Weaken `Sentinel`
* Strip the `Weak` prefix from `WeakIterator` / `WeakInputIterator` / `WeakOutputIterator`

resulting in a smaller overall body of concepts without "boilerplate" constraints. (Some of this is speculative, I haven't done the legwork yet to be certain all the above are possible without breaking something in the design.)

**Comment by ericniebler** [comment]
But this would be the only public place in the design where we require op== without requiring Common. I've been resisting that. If we defined a constrained op== in namespace std for weak (in|out)put Iterators, would that solve the problem?

**Comment by CaseyCarter** [comment]
If the problem was only the hassle of defining the operator, I would let it go: we'll eventually have a tool like `ranges::basic_iterator` that can generate it for us. Similarly, it wouldn't take more than a half hour to define a specialization of `common_type` for `WeakIterator` and `WeakSentinel` that makes them seem to be `EqualityComparable`. What I see as being the true problem here is that the definition space of the predicate is so small: `==` for `WeakIterator`s - and non-weak `Sentinel`s for that matter - effectively has a precondition that the arguments are equal. The compiler should warn you that you are either wasting your time by writing a call to `==` or - more likely - commiting an error. The best way to have the compiler diagnose that error is to not define `==` in the first place.

**Comment by ericniebler** [comment]
IMO, this deserves its own paper.
